### PR TITLE
Issue/155: Timeout while trying to contact RemoteTestNG

### PR DESCRIPTION
**Pull request by missedone**
along with PR https://github.com/cbeust/testng/pull/782 which added while loop for socket.accept, 
this the correponding fix at testng eclipse plugin side. it should fix issue #144 and #155 

when fix this, we have two thing to conside:
1. avoid timeout
2. avoid GUI hang

https://github.com/cbeust/testng/pull/782 takes care of avoiding timeout by polling until the first client connect requrest received.

at plugin side, it puts the initReceive (which has while loop inside) into a thread, here the previous contributor already made it into eclipse Job
the other thing is, plugin need stop the while loop carefully by hook into eclipse launch lifecycle that stop it when the luanch terminate.

### CODE REVIEW

File: testng-eclipse-plugin/src/main/org/testng/eclipse/ui/TestRunnerViewPart.java [hunk 1]
```diff
@@ -8,7 +8,6 @@
 import java.util.List;
 import java.util.Set;
 
-import org.eclipse.core.internal.resources.Workspace;
```
Comment (missedone) [inline testng-eclipse-plugin/src/main/org/testng/eclipse/ui/TestRunnerViewPart.java]: don't use internal api

File: testng-eclipse-plugin/src/main/org/testng/eclipse/ui/TestRunnerViewPart.java [hunk 2]
```diff
@@ -372,23 +378,53 @@ public void run() {
...
                   boolean useProjectJar =
                       TestNGPlugin.getPluginPreferenceStore().getUseProjectJar(project.getProject().getName());
                   String suggestion = useProjectJar
-                     ? "Uncheck the 'Use Project testng.jar' option from your Project properties and try again."
-                     : "Make sure you don't have an older version of testng.jar on your class path.";
-                  new ErrorDialog(m_counterComposite.getShell(), "Couldn't launch TestNG",
-                      "Couldn't contact the RemoteTestNG client. " + suggestion,
-                      new StatusInfo(IStatus.ERROR, "Timeout while trying to contact RemoteTestNG."),
-                      IStatus.ERROR).open();
+                      ? ResourceUtil.getString("TestRunnerViewPart.message.suggestion1")
```
Comment (missedone) [inline testng-eclipse-plugin/src/main/org/testng/eclipse/ui/TestRunnerViewPart.java]: externalize the messages into properties file

File: testng-eclipse-plugin/src/main/org/testng/eclipse/ui/TestRunnerViewPart.java [hunk 3]
```diff
@@ -372,23 +378,53 @@ public void run() {
...
+                           ResourceUtil.getString("TestRunnerViewPart.message.reason")),
+                       IStatus.ERROR).open();
                 }
               });
             }
             return Status.OK_STATUS;
           }
+
+          @Override
+          protected void canceling() {
```
Comment (missedone) [inline testng-eclipse-plugin/src/main/org/testng/eclipse/ui/TestRunnerViewPart.java]: people can cancel the starting job on the process view, so here is hooking into job cancel mechanism, stop the the receiver (aka, break from the while loop)

File: testng-eclipse-plugin/src/main/org/testng/eclipse/ui/TestRunnerViewPart.java [hunk 4]
```diff
@@ -372,23 +378,53 @@ public void run() {
...
+        // handle the process terminated event
+        IDebugEventSetListener listener = new IDebugEventSetListener() {
+          public void handleDebugEvents(DebugEvent[] events) {
+            for (DebugEvent event : events) {
+              if (event.getKind() == DebugEvent.TERMINATE) {
+                Object source = event.getSource();
+                if (source instanceof IProcess) {
+                  IProcess process = (IProcess) source;
+                  if (process.getLaunch().equals(launch)) {
+                    testRunListeningJob.cancel();
```
Comment (missedone) [inline testng-eclipse-plugin/src/main/org/testng/eclipse/ui/TestRunnerViewPart.java]: hook into eclipse launch lifecycle, cancel the job if the launch terminated (e.g. the process end normally, or people killed it)

**Comment by missedone** [comment]
i also updated the testng.jar with the latest snapshot which have new IMessageSender.stopReceiver() interface, so that travis-ci build can pass

File: testng-eclipse-plugin/src/main/org/testng/eclipse/TestNGMessages.properties [hunk 5]
```diff
@@ -27,7 +27,7 @@ TestNGMainTab.projectdialog.title=Project Selection
 TestNGMainTab.projectdialog.message=Choose a project to constrain the search for tests:
 
 TestNGMainTab.runtime.type=Runtime
-TestNGMainTab.testng.compliance=Annotations compliance level (test sources needed for JDK 1.4) 
+TestNGMainTab.testng.compliance=Annotations compliance level (test sources needed for JDK 1.4)
 TestNGMainTab.testng.loglevel=Log level (0-10)
```
Comment (cbeust) [inline testng-eclipse-plugin/src/main/org/testng/eclipse/TestNGMessages.properties]: We should probably get rid of 1.4-related features at some point :)

File: testng-eclipse-plugin/src/main/org/testng/eclipse/ui/TestRunnerViewPart.java [hunk 6]
```diff
@@ -372,23 +378,53 @@ public void run() {
...
                 }
               });
             }
             return Status.OK_STATUS;
           }
+
+          @Override
+          protected void canceling() {
+            if (initialized == false) {
+              TestNGPlugin.log("stop the socket server as it's still pending on accpeting the client");
```
Comment (cbeust) [inline testng-eclipse-plugin/src/main/org/testng/eclipse/ui/TestRunnerViewPart.java]: Typo "accpeting"

**Comment by missedone** [comment]
typo fixed, and remove all unused complianceLevel code

**Comment by cbeust** [comment]
Removing the compliance level would have been fine in a separate PR but thanks :)

**Comment by missedone** [comment]
@cbeust 
could you build a new snapshot of testng eclipse plugin so that we can ask people help verify #144 and #155
thanks

**Comment by cbeust** [comment]
@missedone Done:    6.9.5.201508210528

Let me know how it works for you!

**Comment by missedone** [comment]
@cbeust, just tried, it works for me

**Pull request merged**